### PR TITLE
P6 protocol update

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Address https://github.com/Concordium/concordium-node/issues/856

## Changes

- Introduce migration parameters for P5->P6 protocol update. 
In particular `StateMigrationParametersP5ToP6` is added to the `StateMigrationParameters` which yields the  `StateMigrationData` for protocol 6.

The `StateMigrationData` holds onto the chain parameters that must be supplied for the protocol update P5->P6.

- Consensus parameters for the new chain parameters version V2. These parameters are related to the new consensus protocol.
- Finalization committee parameters. The parameters that determines size and required stake of the finalization committee used in CPV2.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by eb-concordium — APPROVED** [review]

**Review by abizjak — APPROVED** [review]

File: haskell-src/Concordium/Genesis/Data/P6.hs [hunk 1]
```diff
@@ -13,6 +14,47 @@ import qualified Concordium.Crypto.SHA256 as Hash
...
+-- |Parameters used to migrate state from 'P5' to 'P6'.
+data StateMigrationData = StateMigrationData
+    { -- |Data provided by the protocol update to be used
+      -- in the migration.
+      migrationProtocolUpdateData :: !ProtocolUpdateData,
+      -- |The time of the trigger block that caused
+      -- this protocol update.
+      migrationTriggerBlockTime :: !Timestamp
+    }
+    deriving stock (Eq, Show)
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply haskell-src/Concordium/Genesis/Data/P6.hs:48]: It's not. I was just being explicit. Removed

File: haskell-src/Concordium/Genesis/Data/P6.hs [hunk 2]
```diff
@@ -30,8 +72,16 @@ data GenesisDataP6 = GDP6Initial
...
+--
+-- There are two variants, one when migrating from 'P5' to 'P6'and
+-- one from 'P6' to 'P6'.
+data RegenesisP6
+    = GDP6Regenesis {genesisRegenesis :: !BaseV1.RegenesisDataV1}
+    | GDP6RegenesisFromP5
+        { genesisRegenesis :: !BaseV1.RegenesisDataV1,
+          genesisMigration :: !StateMigrationData
+        }
+    deriving stock (Eq, Show)
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply haskell-src/Concordium/Genesis/Data/P6.hs:84]: It's not. I was just being explicit. Removed

File: haskell-src/Concordium/Genesis/Data/P6.hs [hunk 3]
```diff
@@ -102,10 +156,21 @@ regenesisBlockHash GDP6Regenesis{genesisRegenesis = BaseV1.RegenesisDataV1{..}}
     put genesisPreviousGenesis
     put genesisTerminalBlock
     put genesisStateHash
+regenesisBlockHash GDP6RegenesisFromP5{genesisRegenesis = BaseV1.RegenesisDataV1{..}, ..} = BlockHash . Hash.hashLazy . runPutLazy $ do
+    put genesisSlot
+    put P6
+    putWord8 2 -- regenesis variant
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply haskell-src/Concordium/Genesis/Data/P6.hs]: True. I reworded it a bit :) .

File: haskell-src/Concordium/Genesis/Data/P6.hs [hunk 4]
```diff
@@ -30,8 +74,16 @@ data GenesisDataP6 = GDP6Initial
...
+--
+-- There are two variants, one when migrating from 'P5' to 'P6'and
+-- one from 'P6' to 'P6'.
+data RegenesisP6
+    = GDP6Regenesis {genesisRegenesis :: !BaseV1.RegenesisDataV1}
+    | GDP6RegenesisFromP5
+        { genesisRegenesis :: !BaseV1.RegenesisDataV1,
+          genesisMigration :: !StateMigrationData
+        }
+    deriving stock (Eq, Show)
```
Comment (td202) [inline haskell-src/Concordium/Genesis/Data/P6.hs:86]: I don't think it is useful to specify the `stock` deriving strategy here, or in general.

**Review by td202 — APPROVED** [review]
Looks good.

**Pull request merged**